File: solutions/haiku-1/haiku/c/calc.c
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>

typedef struct {
    const char *input;
    int pos;
} Parser;

Parser* parser_create(const char *input) {
    Parser *p = malloc(sizeof(Parser));
    p->input = input;
    p->pos = 0;
    return p;
}

void parser_free(Parser *p) {
    free(p);
}

void skip_whitespace(Parser *p) {
    while (isspace((unsigned char)p->input[p->pos])) {
        p->pos++;
    }
}

int parse_number(Parser *p) {
    skip_whitespace(p);
    int num = 0;
    while (isdigit((unsigned char)p->input[p->pos])) {
        num = num * 10 + (p->input[p->pos] - '0');
        p->pos++;
    }
    return num;
}

int parse_expression(Parser *p);

int parse_primary(Parser *p) {
    skip_whitespace(p);
    if (p->input[p->pos] == '(') {
        p->pos++; // consume '('
        int result = parse_expression(p);
        skip_whitespace(p);
        p->pos++; // consume ')'
        return result;
    }
    return parse_number(p);
}

int parse_factor(Parser *p) {
    int left = parse_primary(p);
    while (1) {
        skip_whitespace(p);
        char op = p->input[p->pos];
        if (op == '*' || op == '/') {
            p->pos++;
            int right = parse_primary(p);
            if (op == '*') {
                left = left * right;
            } else {
                left = left / right;
            }
        } else {
            break;
        }
    }
    return left;
}

int parse_expression(Parser *p) {
    int left = parse_factor(p);
    while (1) {
        skip_whitespace(p);
        char op = p->input[p->pos];
        if (op == '+' || op == '-') {
            p->pos++;
            int right = parse_factor(p);
            if (op == '+') {
                left = left + right;
            } else {
                left = left - right;
            }
        } else {
            break;
        }
    }
    return left;
}

int main() {
    char line[1024];
    while (fgets(line, sizeof(line), stdin)) {
        // Remove newline
        line[strcspn(line, "\n")] = 0;

        Parser *p = parser_create(line);
        int result = parse_expression(p);
        printf("%d\n", result);
        parser_free(p);
    }
    return 0;
}

File: solutions/haiku-1/haiku/javascript/calc.js
const readline = require('readline');

class Parser {
  constructor(input) {
    this.input = input.trim();
    this.pos = 0;
  }

  skipWhitespace() {
    while (this.pos < this.input.length && this.input[this.pos] === ' ') {
      this.pos++;
    }
  }

  peek() {
    this.skipWhitespace();
    if (this.pos < this.input.length) {
      return this.input[this.pos];
    }
    return null;
  }

  parseNumber() {
    this.skipWhitespace();
    let num = 0;
    while (this.pos < this.input.length && /\d/.test(this.input[this.pos])) {
      num = num * 10 + parseInt(this.input[this.pos], 10);
      this.pos++;
    }
    return num;
  }

  parsePrimary() {
    this.skipWhitespace();
    if (this.peek() === '(') {
      this.pos++; // consume '('
      const result = this.parseExpression();
      this.skipWhitespace();
      this.pos++; // consume ')'
      return result;
    }
    return this.parseNumber();
  }

  parseMultDiv() {
    let result = this.parsePrimary();
    while (true) {
      const op = this.peek();
      if (op === '*' || op === '/') {
        this.pos++; // consume operator
        const right = this.parsePrimary();
        if (op === '*') {
          result = result * right;
        } else {
          // Integer division truncating toward zero
          result = Math.trunc(result / right);
        }
      } else {
        break;
      }
    }
    return result;
  }

  parseAddSub() {
    let result = this.parseMultDiv();
    while (true) {
      const op = this.peek();
      if (op === '+' || op === '-') {
        this.pos++; // consume operator
        const right = this.parseMultDiv();
        if (op === '+') {
          result = result + right;
        } else {
          result = result - right;
        }
      } else {
        break;
      }
    }
    return result;
  }

  parseExpression() {
    return this.parseAddSub();
  }

  parse() {
    return this.parseExpression();
  }
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
  terminal: false
});

rl.on('line', (line) => {
  if (line.trim()) {
    const parser = new Parser(line);
    const result = parser.parse();
    console.log(result);
  }
});
